File: time-acceptance/ltFit.cc
#include <iostream>
#include <iomanip>
#include <cstdio>
#include <cassert>
#include <string>

#include <TTree.h>
#include <TFile.h>
#include <TCanvas.h>
#include <TPad.h>
// #include <TSystem.h>
// #include <TCollection.h>

#include <RooGlobalFunc.h>
#include <RooPlot.h>
#include <RooWorkspace.h>
#include <RooFitResult.h>

#include <RooRealVar.h>
#include <RooRealConstant.h>
#include <RooExponential.h>
#include <RooDecay.h>
#include <RooGenericPdf.h>
#include <RooEffProd.h>
#include <RooFormulaVar.h>
#include <RooAddPdf.h>
#include <RooProdPdf.h>
#include <RooDataSet.h>

// #include "utils.hh"

using namespace RooFit;


void ltFit() //TTree *ftree)
{
  TFile rfile("smalltree.root", "read");
  TTree *ftree = (TTree*) rfile.Get("ftree");

  // acceptance
  RooRealVar tau("tau", "lifetime in ns", 0.0001, 0.01);
  RooRealVar turnon("turnon", "turnon", 500, 5000);

  // trigger:
  // HLT2Topo4Body
  // HLT2Topo3Body
  // HLT2Topo2Body
  // HLT2TopoIncPhi
  std::string trigger("HLT2Topo4Body");
  RooRealVar HLT2Topo4Body(trigger.c_str(), trigger.c_str(), 0, 2);

  // dataset
  std::string cut(trigger+">0");
  RooDataSet dataset("dataset", "Dataset", RooArgSet(tau, HLT2Topo4Body),
		     Import(*ftree), Cut(cut.c_str()));

  // when using weights
  RooRealVar wt("wt", "wt", 0, 1e5);
  // weighted dataset
  RooDataSet wdataset("wdataset", "Weighted dataset", RooArgSet(tau, wt, HLT2Topo4Body),
		      WeightVar(wt), Import(*ftree), Cut(cut.c_str()));

  // Decay function: exp(-t*1e3/1.472)
  // RooExponential decay("decay", "Decay function for the B_{s}", tau, // (1)
  // 		       RooRealConstant::value(-1E3/1.472));
  RooExponential decayH("decayH", "Decay function for the B_{s,H}", tau,
			RooRealConstant::value(-1E3/1.536875));
  RooExponential decayL("decayL", "Decay function for the B_{s,L}", tau,
			RooRealConstant::value(-1E3/1.407125));
  RooAddPdf decay("decay", "Decay function for the B_{s}", RooArgList(decayH, decayL),
		  RooRealConstant::value(0.5)); // (2)
  // RooRealVar ratio("ratio", "ration", 0, 1);
  // RooAddPdf decay("decay", "Decay function for the B_{s}", RooArgList(decayH, decayL),
  // 		  RooArgList(ratio));

  // acceptance model: 1-1/(1+(at)³)
  // NB: acceptance is not a PDF by nature
  RooFormulaVar acceptance("acceptance", "1-1/(1+(@0*@1)**3)", RooArgList(turnon, tau)); // for fit
  RooGenericPdf acceptancePdf("acceptancePdf", "@0", RooArgList(acceptance)); // for plot

  RooEffProd Model("Model", "Acceptance model", decay, acceptance);
  Model.fitTo(dataset, Range(0.0001, 0.01)); // SumW2Error(kTRUE), Strategy(2),
  // RooFitResult *fitptr = Model.fitTo(dataset, Range(0.0001, 0.01), Save(true)); // SumW2Error(kTRUE), Strategy(2),

  RooPlot *tframe1 = tau.frame(Name("pfit"), Title("Lifetime acceptance with Monte Carlo"));
  dataset.plotOn(tframe1, MarkerStyle(kFullTriangleUp));
  Model  .plotOn(tframe1);

  RooPlot *tframe2 = tau.frame(Name("pmodel"), Title("a(t) = decay(t) #times acc(t)"));
  wdataset  .plotOn(tframe2, MarkerStyle(kFullTriangleUp)); // , RefreshNorm()
  decay     .plotOn(tframe2, LineColor(kRed));
  // decay     .plotOn(tframe2, LineColor(kRed-9), LineStyle(2), Components(decayL));
  // decay     .plotOn(tframe2, LineColor(kRed-1), LineStyle(2), Components(decayH));
  acceptancePdf.plotOn(tframe2, LineColor(kGreen));
  Model     .plotOn(tframe2, LineColor(kAzure));

  TCanvas *canvas = new TCanvas("canvas", "canvas", 960, 400);
  canvas ->Divide(2,1);
  canvas ->cd(1);
  tframe1->Draw();
  canvas ->cd(2);
  tframe2->Draw();

  std::string ofile(trigger+"_ltFit.pdf");
  canvas ->Print(ofile.c_str());
  canvas ->Print(".png");

  // for (unsigned int i = 0; i < 3; ++i) {
  //   TIter primitivesItr(canvas.cd(i)->GetListOfPrimitives());
  //   while (TObject *obj = primitivesItr()) {
  //     TClass *pclass = TClass::GetClass(obj->ClassName());
  //     std::cout << "Class: " << pclass->ClassName()
  // 	       << " Name: " << obj->GetName()
  // 	       << " Title: " << obj->GetTitle() << std::endl;
  //   }
  // }

  return;
}

File: time-acceptance/ltFit.py
#!/usr/bin/env python

"""
Use RooFit to fit for lifetime acceptance in Monte Carlo
"""

import sys
epsilon = sys.float_info.epsilon

# ROOT classes
from ROOT import TTree, TFile, TCanvas, TPad
# RooFit utilities and ROOT basic types
from ROOT import kGreen, kRed, kBlack, kBlue, kAzure, kYellow, kFullTriangleUp
from ROOT import RooFit, RooGlobalFunc, RooPlot, RooWorkspace, RooFitResult
from ROOT import RooRealVar, RooRealConstant, RooFormulaVar, RooArgSet, RooArgList
from ROOT import RooGenericPdf, RooEffProd, RooAddPdf, RooProdPdf, RooDataSet
from ROOT import RooExponential, RooDecay, RooGaussModel, RooAbsReal

# Get tree
ffile           = TFile('smalltree.root', 'read')
ftree           = ffile.Get('ftree')

# # some RooFit foo for more stable integration
# RooAbsReal.defaultIntegratorConfig().method1D().setLabel("RooAdaptiveGaussKronrodIntegrator1D");
# RooAbsReal.defaultIntegratorConfig().getConfigSection("RooAdaptiveGaussKronrodIntegrator1D").setRealValue("maxSeg", 50);
# RooAbsReal.defaultIntegratorConfig().getConfigSection("RooAdaptiveGaussKronrodIntegrator1D").setCatLabel("method", "31Points");
# RooAbsReal.defaultIntegratorConfig().getConfigSection("RooIntegrator1D").setRealValue("maxSteps", 50);

# observables
time            = RooRealVar('time', 'B_{s} lifetime in ns', epsilon, 0.01)
dt              = RooRealVar('dt', 'Error in lifetime measurement (ns)', epsilon, 1E-4) # limits determined from tree
dt.setBins(100)

# parameters
turnon          = RooRealVar('turnon', 'turnon'    , 500, 5000)
exponent        = RooRealVar('exponent', 'exponent', 2, 4)

# trigger
trigger         = 'HLT2Topo3BodyTOS'
triggerVar      = RooRealVar(trigger, trigger, 0, 2)

# dataset
cut             = trigger+'>0' # hack to avoid entries with negative errors
dataset         = RooDataSet('dataset', 'Dataset',
                             RooArgSet(time, dt, triggerVar),
                             RooFit.Import(ftree), RooFit.Cut(cut))

# weighted dataset
wt              = RooRealVar('wt', 'wt', 0, 1e5)
wdataset        = RooDataSet('wdataset', 'Weighted dataset',
                             RooArgSet(time, wt, triggerVar),
                             RooFit.WeightVar(wt), RooFit.Import(ftree),
                             RooFit.Cut(cut))

# # exponential decay
# decayH          = RooExponential('decayH', 'Decay function for the B_{s,H}',
#                                  time, RooRealConstant.value(-1E3/1.536875))
# decayL          = RooExponential('decayL', 'Decay function for the B_{s,L}',
#                                  time, RooRealConstant.value(-1E3/1.407125))
# decay           = RooAddPdf('decay', 'Decay function for the B_{s}',
#                             decayH, decayL, RooRealConstant.value(0.5))

# resolution model
scale           = RooRealVar('scale', 'Scale factor for lifetime per-event error', 1)
resmodel        = RooGaussModel('resmodel', 'Time resolution model', time,
                                RooRealConstant.value(0), scale, dt) # RooRealConstant.value(0.00004)) # 

# decay model
decayH          = RooDecay('decayH', 'Decay function for the B_{s,H}', time,
                           RooRealConstant.value(1.536875/1E3), resmodel,
                           RooDecay.SingleSided)
decayL          = RooDecay('decayL', 'Decay function for the B_{s,L}', time,
                           RooRealConstant.value(1.407125/1E3), resmodel,
                           RooDecay.SingleSided)
decay           = RooAddPdf('decay', 'Decay function for the B_{s}', decayH, decayL,
                            RooRealConstant.value(0.5))

# acceptance
acceptance      = RooFormulaVar('acceptance', '1-1/(1+(@0*@1)**@2)',
                                RooArgList(turnon, time, exponent))
acceptancePdf   = RooGenericPdf('acceptancePdf', '@0', RooArgList(acceptance))

# Define PDF and fit
Model           = RooEffProd('Model', 'Acceptance model', decay, acceptance)
Model.fitTo(dataset, RooFit.Range(epsilon, 0.01), RooFit.ConditionalObservables(RooArgSet(dt)))

tframe1         = time.frame(RooFit.Name('pfit'), RooFit.Title('Lifetime acceptance with Monte Carlo'))
dataset.plotOn(tframe1, RooFit.MarkerStyle(kFullTriangleUp))
Model  .plotOn(tframe1, RooFit.ProjWData(RooArgSet(dt), dataset, True), RooFit.LineColor(kBlue))

tframe2         = time.frame(RooFit.Name('pmodel'), RooFit.Title('a(t) = decay(t) #times acc(t)'))
wdataset     .plotOn(tframe2, RooFit.MarkerStyle(kFullTriangleUp))
decay        .plotOn(tframe2, RooFit.LineColor(kRed))
Model        .plotOn(tframe2, RooFit.LineColor(kAzure))
acceptancePdf.plotOn(tframe2, RooFit.LineColor(kGreen))

canvas          = TCanvas('canvas', 'canvas', 960, 400)
canvas .Divide(2,1)
canvas .cd(1)
tframe1.Draw()
canvas .cd(2)
tframe2.Draw()
canvas.Print(trigger+'_ltFit_py.png')
